File: homework/Miller-Rabin/MR.py
import random
from Crypto.Util.number import *
import time
def Miller_Rabin(n):
    if n == 2 or n == 3:
        print(f"{n} 是素数")
        return 1
    if n % 2 == 0:
        print(f"{n} 是合数,并且是偶数")
        return 0
    l = 0
    m = n - 1
    while(m%2==0):
        m //= 2
        l += 1
    a = random.randint(2,n-2)
    b = pow(a,m,n)
    if b == 1:
        print(f"{n} 是素数")
        return 1
    for i in range(0,l):
        if b == n-1:
            print(f"{n} 是素数")
            return 1
        else:
            b = pow(b,2,n)
    print(f"{n} 是合数")
    return 0
def test(N):
    start = time.time()
    Miller_Rabin(N)
    end = time.time()
    return (end - start)*1000
N = random.randint(2**2047,2**2048-1)
P = getPrime(2048)
N1 = random.randint(2**4095,2**4096-1)
P1 = getPrime(4096)
print(f"测试2048bit的数")
print(f"{test(N)} ms")
print(f"测试2048bit素数")
print(f"{test(P)} ms")
print(f"测试4096bit的数")
print(f"{test(N1)} ms")
print(f"测试4096bit素数")
print(f"{test(P1)} ms")

File: homework/Miller-Rabin/mr.cpp
#include <NTL/ZZ.h>
#include <iostream>
#include <chrono>
using namespace std;
using namespace NTL;
bool MillerRabin(ZZ n) {
    ZZ m = n - 1;
    int l = 0;
    while (m % 2 == 0) {
        m /= 2;
        l++;
    }
    ZZ a = RandomBnd(n-3) + 2;
    ZZ b = PowerMod(a, m, n);
    if (b == 1) {
        //printf("素数\n");
        return true;
    }
    for(int i = 0;i<l;i++) {
        if (b == n - 1) {
            //printf("素数\n");
            return true;
        }
        b = PowerMod(b, 2, n);
    }
    //printf("合数\n");
    return false;

}
void test(ZZ n) {
    auto start = chrono::high_resolution_clock::now();
    bool result = MillerRabin(n);
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    cout << "Miller-Rabin算法判定" << n << "为" << (result ? "素数" : "合数") << " 耗时: " << duration.count() << " ms" << endl;
}
int main() {
    printf("测试2048bit的素数\n");
    ZZ P = GenPrime_ZZ(2048);
    test(P);
    printf("测试2048bit的数\n");
    ZZ N = RandomBits_ZZ(2048);
    test(N);
    printf("测试4096bit的素数\n");
    ZZ P1 = GenPrime_ZZ(4096);
    test(P1);
    printf("测试4096bit的数\n");
    ZZ N1 = RandomBits_ZZ(4096);
    test(N1);
    return 0;
}